lib/action_controller/serialization.rb

```diff
@@ -25,6 +25,10 @@ def get_serializer(resource, options = {})
           "Please pass 'adapter: false' or see ActiveSupport::SerializableResource#serialize"
         options[:adapter] = false
       end
+      if options[:pagination]
+        options[:original_url] = original_url
+        options[:query_parameters] = query_parameters
+      end
```

Comment:
Just thinking about this part of the pagination interface and how someone might want to generate this outside of a controller

 might we want to instead set these as:

```ruby
options[:pagination][:original_url] ||= original_url
options[:pagination][:query_parameters] ||= query_parameters
```

that way options[:pagination] is still truthy and its options are scoped within it...  and is also passed to the adapter as such

On the other hand, since original url is really a fallback for `options[:links][:self]`, maybe this should be
`options[:links][:self] ||= original_url` (which, now that I think about it, we probably only want to strip the params off the original url for building the non-self links, right?)

Another alternative is to pass `render json: @object, links: { pagination: true }` and we just check if `options[:links].delete(:pagination)` is true. That way, there's one less option being passed around, and pagination piggy backs on links, which it has to do anyways.

I rearranged your code as below while thinking this through.  Hope it's helpful.

```ruby
FIRST_PAGE = 1

def options
  @options ||=
    begin
      {
        pagination: {
          original_url: request.original_url,
          query_parameters: request.query_parameters
        }
      }
    end
end

def links
  @links ||= options[:links] || {}
end

def build_pagination_links(collection)
  pagination_links = {}
  return pagination_links if collection.total_pages == FIRST_PAGE


  original_url = options[:pagination][:original_url]
  url_base = original_url[/\A[^?]+/]
  query_parameters = options[:query_parameters] || original_url[/\?.*/] || {}

  current_links = links
  self_link = current_links.fetch(:self) { original_url }

  collection_size            = collection.size
  collection_current_page    = collection.current_page
  collection_total_pages     = collection.total_pages

  link_template = "#{url_base}?%s"
  if collection_current_page != FIRST_PAGE
    pagination_links[:first] = link_template % query_parameters.merge(size: collection_size, number: FIRST_PAGE).to_query
    pagination_links[:prev]  = link_template % query_parameters.merge(size: collection_size, number: (collection_current_page - FIRST_PAGE)).to_query
  end

  if collection_current_page != collection_total_pages
    pagination_links[:next] = link_template % query_parameters.merge(size: collection_size, number: (collection_current_page + FIRST_PAGE)).to_query
    pagination_links[:last] = link_template % query_parameters.merge(size: collection_size, number: collection_total_pages).to_query
  end
end

links.update(build_pagination_links(paginated_resource))
```